contracts/src/LightClient.sol

```diff
@@ -373,6 +386,35 @@ contract LightClient is Initializable, OwnableUpgradeable, UUPSUpgradeable {
         }
     }
 
+    /// @notice update the stateUpdateBlockNumbers and hotShotCommitments arrays each time a new
+    /// finalized state is added
+    /// @dev only maintains maxStateHistoryAllowed non-zero elements in these arrays
+    /// @dev using a FIFO approach to delete elements at the start of the array and storing the new
+    /// initial index in stateHistoryFirstIndex
+    function updateStateHistory(uint256 blockNumber, LightClientState memory state) internal {
+        if (maxStateHistoryAllowed == 0) revert InvalidMaxStateHistory();
+
+        if (stateHistoryCount == maxStateHistoryAllowed) {
```

Comment:
The `maxStateHistoryAllowed` name is a bit misleading right now because if the setter `setMaxStateHIstoryAllowed` is used to reduce it then size of the history that is currently larger it won't be reduced because we only ever remove an element if we also add one. Unless I'm misunderstanding something.

Reply:
I really doubt we ever want to set a smaller state history after deployment though so because all that it allows us to do is reclaim some gas. So maybe we can "fix it" by not allowing setting a lower state history?

We could also have functionality to prune multiple entries but the problem with that is that we need to then have potential issues where calls can run out of gas if the iteration is too long. 

One way to resolve this is also to allow for a few (at least 2) pruning operations during an update. This makes the gas usage bounded and still allows the history to shrink.

Reply:
your understanding is correct, it does not considering the reduction of the max state history allowed. Would you allow me to add the ability to prune +1 elements at the same time? and if so, I can make it so that the maxStateHistoryAllowed can be reduced. Otherwise, i can restrict modifications to only increasing the max state history allowed

Reply:
I think any changes here are somewhat problematic because it potentially affects the result of challenges in the rollups. Increasing the history size definitely seems less problematic though.

Reply:
ah true - also this light client may be deployed on L2s to support L3s - Would there be a need to shorten the `maxStateHistoryAllowed` in those instances?

Reply:
Currently we hardcode the `maxStateHistoryAllowed` in the `_initializeState` method but considering that the light client would also deploy on L2s to support L3s. Should it be configurable at deployment?

Reply:
L2s are also a lot cheaper so I wonder if it would be acceptable to default to not prune the history on L2s.

Reply:
Another alternative would be to use `block.timestamp` instead of block numbers. That way we could at least set a default that should keep working on L2s as long as their timestamps are correct. I'm currently not 100% clear on how trustworthy `block.timestamp` is, especially on L2s. I would hope that validators would only tolerate small deviations though.

Reply:
Based on the answer here: it sounds like the timestamp is pretty safe to use after the merge. Wdyt @alysiahuggins ?

https://ethereum.stackexchange.com/a/140818

Reply:
this has been addressed in a previous commit